include/mbgl/actor/actor.hpp

```diff
@@ -8,9 +8,108 @@
 #include <memory>
 #include <future>
 #include <type_traits>
+#include <cassert>
 
 namespace mbgl {
 
+template <class Object>
+class EstablishedActor;
+
+template <class Object>
+class Actor;
+
+template <class Object>
+class AspiringActor {
+public:
+    AspiringActor() : mailbox(std::make_shared<Mailbox>()) {
+        assert(!mailbox->isOpen());
+    }
+    
+    AspiringActor(const AspiringActor&) = delete;
+    
+    template <class... Args>
+    static std::tuple<std::decay_t<Args>...> captureArguments(Args&&... args) {
+        return std::make_tuple(std::forward<Args>(args)...);
+    }
+
+    ActorRef<std::decay_t<Object>> self() {
+        return ActorRef<std::decay_t<Object>>(object(), mailbox);
+    }
+    
+    template <typename Fn, class... Args>
+    void invoke(Fn fn, Args&&... args) {
```

Comment:
I wonder if we should just require using `a.self().invoke(...)` and `a.self().ask(...)` instead of having a separate pathway for `Actor` vs `ActorRef`.

Reply:
Ah - I assumed we intentionally had different pathways to avoid having to make the ActorRef (with its `weak_ptr`) for each call